fix(proc): ensure nvim_get_proc_children() returns all child processes

Xeron2000 opened the pull request:
Fix #28741
Problem:
The original implementation only checked the main thread's children file (/proc/<ppid>/task/<ppid>/children), missing child processes created by other threads.

Solution:
Now, it iterates through all thread directories under /proc/<ppid>/task/, reads each thread's children file, collects unique child PIDs, and properly handles errors. This ensures complete detection of child processes regardless of which thread created them.
```
echo "
        vim.system{'firefox','-P'}
        vim.uv.sleep(1000)
        local pid=assert(tonumber(vim.iter(vim.split(vim.system{'pgrep','firefox'}:wait().stdout,'\n',{trimempty=true})):nth(-1)))
        vim.print('Child processes: '..vim.inspect(vim._os_proc_children(pid)))
        vim.print('nvim_get_proc_children: '..vim.inspect(vim.api.nvim_get_proc_children(pid)))
        "|/usr/local/bin/nvim -l -
```
output
```
Child processes: { 150113, 150146 }
nvim_get_proc_children: { 150113, 150146 }
```

justinmk commented:
Thanks! Since this is using linux's procfs, could be done in Lua. We already have a Lua function that is called by `nvim_get_proc_children`:  https://github.com/neovim/neovim/blob/8b5a0a00c8cfe776c4227862c3fb32a07d154663/runtime/lua/vim/_editor.lua#L179

Proposal:

- change `os_proc_children` so that it immediately does `return 2` for linux.
- update the Lua function for linux
- (optional followup PR (not this PR)): rearrange the code so that `os_proc_children` internally calls the Lua function, instead of doing that in `nvim_get_proc_children`. This allows `os_proc_children` to be useful on all platforms.

Xeron2000 commented:
### **Summary of Existing Implementation, Issues, and My Thoughts**

- **vim._os_proc_children()** relies on the `pgrep -P` command to retrieve child processes.  
- **os_proc_children()** on Linux reads `/proc/<ppid>/task/<tid>/children` to get child processes, but the current implementation only checks the main thread, which may cause it to miss child processes created by other threads.  
- **vim.api.nvim_get_proc_children()** simply calls `os_proc_children()`, so it is limited by the behavior of `os_proc_children()`.  

The current Linux implementation of `os_proc_children()` can find the main thread, but since child processes may not necessarily be created by the main thread, it fails to return `2` and does not retrieve all child processes.  

You suggested modifying os_proc_children(ppid) to return 2 on Linux, so that vim.api.nvim_get_proc_children(ppid) calls vim._os_proc_children(ppid) in Lua, allowing pgrep to fetch child processes.

However, since the issue originates from `os_proc_children()`, it can be resolved by simply improving the Linux-specific code in `os_proc_children()`.  

### **Fix**

Instead of modifying the Lua logic that uses `pgrep`, we should keep the Linux logic in the C code and enhance it to traverse `/proc` completely. This will ensure multi-threading support and prevent missing child processes.

justinmk commented:
> Instead of modifying the Lua logic that uses `pgrep`, we should keep the Linux logic in the C code and enhance it to traverse `/proc` completely

No, I meant that the linux logic should be moved to Lua, and it can read /proc there. I did not mean that pgrep should be used on linux.

There is no reason to do all of this in C.

Xeron2000 commented:
> > Instead of modifying the Lua logic that uses `pgrep`, we should keep the Linux logic in the C code and enhance it to traverse `/proc` completely
> 
> No, I meant that the linux logic should be moved to Lua, and it can read /proc there. I did not mean that pgrep should be used on linux.
> 
> There is no reason to do a bunch of filesystem operations (traversing /proc) in C.

Got it, I misunderstood. Thanks for the clarification.

justinmk commented:
Haven't looked closely at the logic, but this looks much better! Is there a small meaningful change that can be made to the existing test(s)? Or do we need a new test? `test/functional/api/proc_spec.lua`

You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#filter-by-file) like this:

    TEST_FILE=test/functional/api/proc_spec.lua make functionaltest

For testing timing-sensitive behavior, note that:

- `t.retry()` is also available to retry assertions that are waiting for something.

Xeron2000 committed:
fix(proc): ensure nvim_get_proc_children() returns all child processes

Problem:
The original implementation only checked the main thread's children file
(/proc/<ppid>/task/<ppid>/children), missing child processes created by
other threads.

Solution:
Iterate through all thread directories under /proc/<ppid>/task/, read
each thread's children file, collect unique child PIDs, and properly
handle errors. This ensures complete detection of child processes
regardless of which thread created them.
